### PR TITLE
Fix test panics for submodule of book is not updated

### LINKED ISSUE #113963: `./x.py test` panics
```
$ ./x.py test
Building bootstrap
    Finished dev [unoptimized] target(s) in 0.18s
thread 'main' panicked at 'fs::read_dir(builder.src.join(&relative_path).join("redirects")) failed with No such file or directory (os error 2)', doc.rs:228:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:00:01
```
This might not be a valid request (I have to tell it what to test), but surely it shouldn't just result in such an obscure panic.

Cc @rust-lang/bootstrap

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #113963

### CODE REVIEW

**Comment by ozkanonur** [comment]
This will not give the error(which it should give) if submodule was fetched but directory was corrupted/removed. Checking if submodule was fetched or not could be better option here.

**Comment by chenyukang** [comment]
updated the code to check the directory and exit(1) if checking failed.

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -197,11 +197,21 @@ impl Step for TheBook {
         let compiler = self.compiler;
         let target = self.target;
 
+        let absolute_path = builder.src.join(&relative_path);
+        let redirect_path = absolute_path.join("redirects");
+        if !absolute_path.exists()
+            || !redirect_path.exists()
+            || dir_is_empty(&absolute_path)
+            || dir_is_empty(&redirect_path)
+        {
```
Comment (ozkanonur) [inline src/bootstrap/doc.rs:206]: We can create function something like `fn is_submodule_checked_out(module_path: &str) -> bool` that uses `git submodule foreach` command inside.

**Review by onur-ozkan — APPROVED** [review]
Thanks for the fix! Looks good to me.

Only 1 suggestion you may want to consider following it. r=me with/without it since it's not a blocker.

**Comment by rylev** [comment]
Perf: this is noise that is corrected in the next run.

@rustbot label: +perf-regression-triaged

**Pull request merged**